### ISSUE TITLE
I will create many pod to run Alluxio in K8S，so do I need apache ratis？

### ISSUE THREAD

**Issue by jiankebe** [Alluxio/alluxio#17332]

**Comment by LuQQiu** [comment]
@ssz1997 Please help with the question?

**Comment by ssz1997** [comment]
@jiankebe Hi thank you for your question. Can you provide more information about what you want to achieve? 

Here're some information that may answer your question:
1. Alluxio master has a raft-based mechanism to do leader election. You don't need extra service to achieve high availability.
2. All other Alluxio components are mostly stateless, so there's no ratis needed.
3. If you have multiple Alluxio clusters, we can discuss more about the communication between them.

**Comment by jiankebe** [comment]
I see that Journal save in local file, but in K8S pod the local directory is not persistent
so I will use pvc to share directory to mapping that directory in order to it is not lost, or Can i use redis(or RocksDB) to save Journal data?
If use RocksDB，how to deploy independently Like Mysql

**Comment by ssz1997** [comment]
@jiankebe Yes journal is saved in local file. We do recommend using embedded journal, in which each master keeps its own journal, so it is not necessary to have a shared directory for all masters. Therefore, local PV or hostPath as volume would work perfectly fine. Adding more services would only add more complexity of your system which I would not recommend.

**Comment by jiankebe** [comment]
> @jiankebe Yes journal is saved in local file. We do recommend using embedded journal, in which each master keeps its own journal, so it is not necessary to have a shared directory for all masters. Therefore, local PV or hostPath as volume would work perfectly fine. Adding more services would only add more complexity of your system which I would not recommend.

Thank you

**Comment by jiankebe** [comment]
> @jiankebe Yes journal is saved in local file. We do recommend using embedded journal, in which each master keeps its own journal, so it is not necessary to have a shared directory for all masters. Therefore, local PV or hostPath as volume would work perfectly fine. Adding more services would only add more complexity of your system which I would not recommend.

As you said，embedded journal is raft logs data。Alluxio use ratis to solve data consistency issues on multi master ？

**Comment by jiankebe** [comment]
> > @jiankebe Yes journal is saved in local file. We do recommend using embedded journal, in which each master keeps its own journal, so it is not necessary to have a shared directory for all masters. Therefore, local PV or hostPath as volume would work perfectly fine. Adding more services would only add more complexity of your system which I would not recommend.
> 
> As you said，embedded journal is raft logs data。Alluxio use ratis to solve data consistency issues on multi master ？

or Alluxio does not support the deployment of multiple master nodes

**Comment by jiankebe** [comment]
> > > @jiankebe Yes journal is saved in local file. We do recommend using embedded journal, in which each master keeps its own journal, so it is not necessary to have a shared directory for all masters. Therefore, local PV or hostPath as volume would work perfectly fine. Adding more services would only add more complexity of your system which I would not recommend.
> > 
> > 
> > As you said，embedded journal is raft logs data。Alluxio use ratis to solve data consistency issues on multi master ？
> 
> or Alluxio does not support the deployment of multiple master nodes

If each host stores data independently, is there a synchronization mechanism between them, and how to keep data consistent

**Comment by ssz1997** [comment]
@jiankebe We support multi-master deployment for high availability of the system.

@jenoudet Do you mind briefly introduce how the journal works? His main questions are 
1. Alluxio use ratis to solve data consistency issues on multi master ？
2. If each host stores data independently, is there a synchronization mechanism between them, and how to keep data consistent

**Comment by jenoudet** [comment]
Hello @jiankebe,
1. Yes, Alluxio does use [Apache Ratis](https://ratis.apache.org/) to solve data consistency issues when using multiple masters.
2. Apache Ratis is a Java implementation of the [Raft Consensus Algorithm](https://raft.github.io/). It uses this algorithm to keep data consistent and synchronized across masters.

**Comment by jiankebe** [comment]
> Hello @jiankebe,
> 
> 1. Yes, Alluxio does use [Apache Ratis](https://ratis.apache.org/) to solve data consistency issues when using multiple masters.
> 2. Apache Ratis is a Java implementation of the [Raft Consensus Algorithm](https://raft.github.io/). It uses this algorithm to keep data consistent and synchronized across masters.

about 2 answer。 Does each master store a full amount of raft logs?  and when using multiple masters Is there only one node as the master to accept requests, and other nodes are as the standby node to not accept requests and only synchronize raft logs when the master node shutdown the standby node will improve status to as master?

**Comment by ssz1997** [comment]
1. Each master has its own copy, but not necessarily full. In other words, they are not synced at all times.
2. Only one master accepts requests. Before client sends out requests, it first asks who is the leading master, and then sends out the request.

**Comment by jiankebe** [comment]
> 1. Each master has its own copy, but not necessarily full. In other words, they are not synced at all times.
> 2. Only one master accepts requests. Before client sends out requests, it first asks who is the leading master, and then sends out the request.

Thanks!!!
